### PR TITLE
Upgrade to > Qt5.15 -6 and ImGUI > v1.87 + fixes keyboard and mouse handling & widget samples

### PR DESCRIPTION
**Pull request by moezb**

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -4,7 +4,18 @@ project(qtimgui)
 set(CMAKE_AUTOMOC ON)
 set(CMAKE_AUTORCC ON)
 set(CMAKE_AUTOUIC ON)
-find_package(Qt5 COMPONENTS Core Quick Gui Widgets REQUIRED)
+
+# We're using qt versionless targets so our minimal Qt version is 5.15, we support Qt6 as well
+# See https://www.qt.io/blog/versionless-cmake-targets-qt-5.15
+find_package(Qt6 COMPONENTS COMPONENTS  Core Quick Gui Widgets OpenGLWidgets )
```
Comment (suy) [inline src/CMakeLists.txt:10]: Extra space before the parenthesis and before "Core".

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -4,7 +4,18 @@ project(qtimgui)
...
+if (NOT Qt6_FOUND)
+    find_package(Qt5 5.15 REQUIRED COMPONENTS Core Quick Gui Widgets)
+    if (NOT Qt5_FOUND)
+        message(FATAL_ERROR "We only support  only Qt 5.15 and higher.")
+    endif()
+endif()
+
+message(STATUS  Qt ${QT_DEFAULT_MAJOR_VERSION})
```
Comment (suy) [inline src/CMakeLists.txt:18]: This message should be a bit clearer to the user. Like "Compiling with Qt X". I saw it, and was just a bit puzzled.

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -21,7 +32,10 @@ target_link_libraries(
     qt_imgui_quick
     PUBLIC
     imgui
-    Qt5::Gui
+    Qt::Gui
+    # We use the new Qt Rendering Hardware Interface (RHI) when buzilding with Qt and thgus we need openglwidgets qt module
```
Comment (suy) [inline src/CMakeLists.txt:35]: s/buzilding with Qt and thgus/building with Qt and thus/

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -34,7 +48,10 @@ target_link_libraries(
     qt_imgui_widgets
     PUBLIC
     imgui
-    Qt5::Widgets
+    Qt::Widgets
+    # We use the new Qt Rendering Hardware Interface (RHI) when buzilding with Qt and thgus we need openglwidgets qt module
```
Comment (suy) [inline src/CMakeLists.txt:51]: Ditto.

File: src/key_mappings.h [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+#ifdef Q_OS_ANDROID
+    #include "platform/android_key_mappings.cpp"
+#endif
+
+#ifdef Q_OS_MACOS
+    #include "platform/macosx_key_mappings.cpp"
+#endif
+
+#ifdef Q_OS_LINUX
+    #include "platform/linux_key_mappings.cpp"
```
Comment (suy) [inline src/key_mappings.h:20]: The one for linux is entirely empty. Is there a recipe one can follow to make the mappings? I checked the backend directory on ImGui, but I was expecting something like X11/Wayland, and that doesn't seem to be the case.

File: examples/widget/demo-widget.cpp [hunk 6]
```diff
@@ -68,22 +81,24 @@ int main(int argc, char *argv[])
...
 
     // Show window
-    DemoWindow w;
-    w.setWindowTitle("QtImGui widget example");
-    w.resize(1280, 720);
-    w.show();
-
+    DemoWidget *widget=new DemoWidget();
+    widget->setWindowTitle("QtImGui widget example");
+    QWidget* window=new QWidget();
```
Comment (suy) [inline examples/widget/demo-widget.cpp:93]: I don't think this should be heap allocated. It seems to be leaking right now.

File: src/ImGuiRenderer.cpp [hunk 7]
```diff
@@ -401,90 +415,112 @@ void ImGuiRenderer::onKeyPressRelease(QKeyEvent *event)
...
-    if(io.WantSetMousePos) {
-        m_window->setCursorPos({(int)io.MousePos.x, (int)io.MousePos.y});
+    ImGui::SetCurrentContext(g_ctx);
+    ImGuiIO& io = ImGui::GetIO();
+
+    if (watched == m_window->object()) {
+        
+        switch (event->type()) {
+        case QEvent::MouseMove:
+            onMouseMove(dynamic_cast<QMouseEvent*>(event));
```
Comment (suy) [inline src/ImGuiRenderer.cpp:480]: This cast, and all of the above, can be `static_cast`.

**Comment by seanchas116** [comment]
Looks like it doesn't compile on some platforms so I merged it to the separate `upgrade-imgui` branch.

**Pull request merged**